lib/OnyxCache.js

```diff
@@ -0,0 +1,183 @@
+     * the value is not available in cache
+     * @template T
+     * @param {string} key
+     * @param {function(key: string): Promise<T>} fallback - a fallback to retrieve the item from storage
+     * @returns {Promise<T>}
+     */
+    getValue(key, fallback) {
+        // When we have cache resolve right away
+        if (this.hasCacheForKey(key)) {
+            return Promise.resolve(this.storageMap[key]);
+        }
+
+        // Get the value using the fallback. This task optimizes concurrent calls see `resolveTask` for details
+        const startTask = () => fallback(key).then(value => this.update(key, value));
+        return this.resolveTask(`getValue:${key}`, startTask);
+    }
+
+    /**
+     * Check whether cache has data for the given key
+     * @param {string} key
+     * @returns {boolean}
+     */
+    hasCacheForKey(key) {
+        return isDefined(this.storageMap[key]);
+    }
+
+    /**
+     * Update or add data to cache
+     * Providing value is optional. Skipping value would just capture the key so that
+     * it's available in `getAllKeys`
+     * @param {string} key
+     * @param {*} [value]
+     * @returns {*} value - returns the cache value
+     */
+    update(key, value) {
+        // Keep storage keys up to date so that `getAllKeys` works correctly
+        this.storageKeys.add(key);
+
+        // When a 2nd parameter (value) is provided - store it under the key
+        if (isDefined(value)) {
+            this.storageMap[key] = value;
+        }
+
+        return value;
+    }
+
+    /**
+     * Remove data from cache
+     * @param {string} key
+     */
+    remove(key) {
+        this.storageKeys.delete(key);
+        delete this.storageMap[key];
+    }
+
+    /**
+     * Merge data to cache, any non existing keys will be crated
+     * Array items are concatenated
+     * Object items are merged
+     * If the existing value in cache and the new value are not the same type
+     * the existing value is replaced by the new value
+     * In case the merge value for a key is undefined - the key is removed from cache
+     * @param {Array<[string, *]>} pairs - array of key value pairs
+     */
+    merge(pairs) {
+        _.forEach(pairs, ([key, valueToMerge]) => {
+            if (_.isUndefined(valueToMerge)) {
+                this.remove(key);
+                return;
+            }
+
+            const existingValue = this.storageMap[key];
+            let newValue = valueToMerge;
+
+            /* Perform merge or concatenation for complex values otherwise just replace the old value */
+
+            if (_.isObject(valueToMerge) && _.isObject(existingValue)) {
+                newValue = {...existingValue, ...valueToMerge};
+            }
+
+            if (_.isArray(valueToMerge) && _.isArray(existingValue)) {
+                newValue = [...existingValue, ...valueToMerge];
+            }
+
+            this.update(key, newValue);
```

Comment:
Here's the same for `AsyncStorage`:

![image](https://user-images.githubusercontent.com/12156624/120030231-dff6ef00-bfff-11eb-98de-b5c53f7d098f.png)

So it looks like the end result of both `AsyncStorage.merge` and `lodashMerge` is identically and we should use `lodashMerge` in cache

Reply:
Yes the cache should be performing the same logic as `multiMerge()` ?

https://github.com/react-native-async-storage/async-storage/blob/864c64bd0f6414c8959548ec237637a82b7faec3/src/AsyncStorage.js#L13-L19

Since you want to avoid changing too much behavior from what we have now the path of least resistance would be to rename this to `cache.multiMerge()` instead of `cache.merge()`.

This is mainly to avoid confusion with `Onyx.merge()` which actually batches merges + handles arrays properly + calls `set()`.

Sound good?

Reply:
I was thinking of just make it a regular merge that will get a single object and merge it to cache

We have only one use case here:

https://github.com/Expensify/react-native-onyx/blob/ee177e27eeae0208d3aeeef451c4eedf8badfe3b/lib/Onyx.js#L647-L648

That will be updated to 

```js
cache.merge(collection);
```

Since `loadashMerge` would update nested objects the delta from the `collection` would be merged to cache and the cache keys for the collection items updated

The other usage here is invalid

https://github.com/Expensify/react-native-onyx/blob/ee177e27eeae0208d3aeeef451c4eedf8badfe3b/lib/Onyx.js#L647-L648

And will be removed and the code will do this instead:

```js
    _.each(data, (val, key) => {
        // Update cache and optimistically inform subscribers
        cache.set(key, val);
        keyChanged(key, val);
    });
```

Reply:
Updated `merge` to work with an object and to use `lodashMerge` internally

AsyncStorage `multiSet`/`multiMerge` uses key value pairs and for each pair `merges` (or sets) the data for the key
`cache.merge` uses a single object with cache keys and values (collection) - it merges data to those keys 
The end result is the same as the stored result